torch_xla/csrc/aten_xla_type.cpp

```diff
@@ -879,16 +879,116 @@ at::Tensor& XLANativeFunctions::arange_out(const at::Scalar& start,
+  // Optimize: decide if we can use `slice` to replace `as_strided` to avoid the
+  // copy. To use `slice`, the following conditions must be satisfied:
+  // - `stride` must be: [..., dim[-2]*dim[-1], dim[-1], 1] and one of them
+  //   (assume stride[X]) can be multiplied with a constant K.
+  // - `size` must be: [dim[0], dim[1], ...], and size[X]*K <= dim[X].
+  // In theory we can shuffle element in `stride` and `size` and this can result
+  //   in the transpose of dimensions, but we don't consider this case here.
+  auto tensor_dim = tensor.sizes();
+  if ((tensor_dim.size() == stride.size() ||
+       tensor_dim.size() == stride.size() + 1) &&
+      (!storage_offset.has_value() || (*storage_offset) == 0)) {
+    // find the dim that we either skip or slice on based on stride
+    long l = tensor_dim.size();
+    long stride_mul = 1;
+    long skip_dim = -1;
+    int K = 0;
+    // check `stride`
+    for (long i = l - 1, j = std::min(i, (long)stride.size() - 1); j >= 0;
+         i--, j--) {
+      if (stride_mul != stride[j]) {
+        if (skip_dim == -1) {
+          skip_dim = i;
+          K = stride[j] / stride_mul;
+          if (tensor_dim.size() == stride.size() + 1) {
+            // tensor_dim and stride element can potentially shift by one.
+            j++;
+          }
+        } else {
+          // multiple X index found
+          return bridge::AtenFromXlaTensor(nullptr);
+        }
+      }
+      stride_mul *= tensor_dim[i];
+    }
+
+    if (tensor_dim.size() == stride.size() + 1) {
+      for (long i = 0, j = 0; i < size.size(); i++, j++) {
+        if (i == skip_dim) {
+          j++;
+        } else {
+          if (size[i] != tensor_dim[j]) {
+            return bridge::AtenFromXlaTensor(nullptr);
+          }
+        }
+      }
+      return bridge::AtenFromXlaTensor(tensor_methods::squeeze(
+          tensor_methods::slice(bridge::GetXlaTensor(tensor), skip_dim, 0, 1,
+                                1),
+          skip_dim));
+    } else if (tensor_dim.size() == stride.size()) {
+      long reduce_size_location = -1;
+      for (auto i = 0; i < l; i++) {
+        if (size[i] != tensor_dim[i]) {
+          if (size[i] < tensor_dim[i] && reduce_size_location == -1) {
+            reduce_size_location = i;
+          } else {
+            return bridge::AtenFromXlaTensor(nullptr);
+          }
+        }
+      }
+      // check if only one dimension is sliced
+      if (reduce_size_location != -1) {
+        if (skip_dim != -1) {
+          if (skip_dim != reduce_size_location ||
+              size[reduce_size_location] * K >
+                  tensor_dim[reduce_size_location]) {
+            return bridge::AtenFromXlaTensor(nullptr);
+          }
+        } else {
+          // we have one dim size reduced but without any step jump regarding
+          // stride.
+          K = 1;
+        }
+      } else {
+        // size remains the same as tensor, we can't return the same tensor
```

Comment:
This can also be rewritten to an early exit style:

```
if (reduce_size_location == -1) {
  // ...
  return at::Tensor();
}
```

In summary, it's clearer to hoist any single line "return ..." code path as early as possible. Then we don't need a deep indentation nor do we need an `else` block. In well written complex C++ logic we'll often see something like this:

```
if (foo) {
  return ...;
}
if (bar) {
  return ...;
}
if (baz) {
  return ...;
}
// Now the real stuff begins
...
```